File: MergeSort.py
#Merge sort

def MergeSort(arr,low,high):
    if low<high:
       mid=(low+high)//2
       MergeSort(arr,low,mid)
       MergeSort(arr,mid+1,high)
       Merge(arr,low,mid,mid+1,high)

def Merge(arr,i1,j1,i2,j2):
    temp=[0]*10000
    i=i1
    j=i2
    k=0
    while(i<=j1 and j<=j2):
        if(arr[i]<arr[j]):
            temp[k]=arr[i];
            k+=1
            i+=1
        else:
            temp[k]=arr[j];
            k+=1
            j+=1
    while(i<=j1):
        temp[k]=arr[i];
        k+=1
        i+=1
    while(j<=j2):    
        temp[k]=arr[j];
        k+=1
        j+=1
    j=0
    for i in range(i1,j2+1):
        arr[i]=temp[j]
        j+=1

arr = list(map(int,input("Enter the elements(seperated by spaces): ").split(' ')))
MergeSort(arr,0,len(arr)-1)
print(f"The elements in sorted order: {arr}")

File: Merge_Sort.cpp
#include<vector>
#include<iostream>

using namespace std;


void merge(vector<int>&arr,int s,int e){
    int i=s;
    int m=s+(e-s)/2;
    int j=m+1;
    vector<int>tmp;
    while ((i<=m) && (j<=e))
    {
        if(arr[i]<=arr[j]){
            tmp.push_back(arr[i++]);
        }
        else{
            tmp.push_back(arr[j++]);
        }
    }
    while(i<=m){
        tmp.push_back(arr[i++]);
    }
    while(j<=e){
        tmp.push_back(arr[j++]);
    }

    int k=0;
    for(int l=s;l<=e;l++){
        arr[l]=tmp[k++];
    }

    return;
    
}

void mergeSort(vector<int>&arr,int s,int e){
    // base case
    if(s>=e) return;
    // recursive
    int m=s+(e-s)/2;
    mergeSort(arr,s,m);     // 10 5 2 0... 10 5..10
    mergeSort(arr,m+1,e);   // 5...2 0...7 6 4
    return merge(arr,s,e);    
}


int main(){

    int n;cin>>n;
    vector<int>arr(n);   // 10 5 2 0 7 6 4
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    
    mergeSort(arr,0,(n-1));

    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }
    cout<<"\n";

}
